kv/kvserver: TestEagerReplication failed

[(kv/kvserver).TestEagerReplication failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2287197&tab=buildLog) on [master@0afb350a2e41bd080af4b67990dc677a2acfbb58](https://github.com/cockroachdb/cockroach/commits/0afb350a2e41bd080af4b67990dc677a2acfbb58):

```
=== RUN   TestEagerReplication
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestEagerReplication900908763
--- FAIL: TestEagerReplication (45.17s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestEagerReplication900908763
    test_log_scope.go:63: use -show-logs to present logs inline
    replicate_test.go:46: purgatory start count is 35
    replicate_test.go:60: condition failed to evaluate within 45s: expected 36 replicas in purgatory, but found 35
        goroutine 1172490 [running]:
        runtime/debug.Stack(0xc0031b9c40, 0x5b70880, 0xc003ac5360)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x9d
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x5d762e0, 0xc002f4c200, 0xc0031b9c40)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:36 +0x6b
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestEagerReplication(0xc002f4c200)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replicate_test.go:60 +0x483
        testing.tRunner(0xc002f4c200, 0x50f9540)
        	/usr/local/go/src/testing/testing.go:909 +0xc9
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:960 +0x350

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestEagerReplication PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestEagerReplication.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestEagerReplication failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2355430&tab=buildLog) on [master@e824da5a2b33168fa4ef93c83295c8205acfadb7](https://github.com/cockroachdb/cockroach/commits/e824da5a2b33168fa4ef93c83295c8205acfadb7):

```
=== RUN   TestEagerReplication
--- FAIL: TestEagerReplication (0.01s)
    test_log_scope.go:107: can't use TestLogScope with secondary loggers active

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=TestEagerReplication PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestEagerReplication.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestEagerReplication failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2760887&tab=buildLog) on [master@dd87d114508b8b34230ae7da560ab053dd4f58ce](https://github.com/cockroachdb/cockroach/commits/dd87d114508b8b34230ae7da560ab053dd4f58ce):

```
=== RUN   TestEagerReplication
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestEagerReplication272306616
    test_log_scope.go:74: use -show-logs to present logs inline
    replicate_test.go:54: purgatory start count is 6
    replicate_test.go:68: condition failed to evaluate within 45s: expected 8 replicas in purgatory, but found 9
        goroutine 1582501 [running]:
        runtime/debug.Stack(0xc00272fb50, 0x5436ac0, 0xc00bdab480)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x9f
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x5615e00, 0xc007c91200, 0xc00272fb50)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:36 +0x6d
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestEagerReplication(0xc007c91200)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replicate_test.go:68 +0x4c5
        testing.tRunner(0xc007c91200, 0x4e21cb0)
        	/usr/local/go/src/testing/testing.go:1123 +0xef
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1168 +0x2b3
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestEagerReplication272306616
--- FAIL: TestEagerReplication (45.50s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestEagerReplication PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestEagerReplication.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestEagerReplication failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2761738&tab=buildLog) on [master@3bae381e45397225ff926f3f8f60d52b70fdc661](https://github.com/cockroachdb/cockroach/commits/3bae381e45397225ff926f3f8f60d52b70fdc661):

```
=== RUN   TestEagerReplication
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestEagerReplication985901282
    test_log_scope.go:74: use -show-logs to present logs inline
    replicate_test.go:54: purgatory start count is 7
    replicate_test.go:68: condition failed to evaluate within 45s: expected 9 replicas in purgatory, but found 10
        goroutine 1620387 [running]:
        runtime/debug.Stack(0xc002429b50, 0x5436be0, 0xc0016e5740)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x9f
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x5615f20, 0xc002021380, 0xc002429b50)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:36 +0x6d
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestEagerReplication(0xc002021380)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replicate_test.go:68 +0x4c5
        testing.tRunner(0xc002021380, 0x4e21cd8)
        	/usr/local/go/src/testing/testing.go:1123 +0xef
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1168 +0x2b3
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestEagerReplication985901282
--- FAIL: TestEagerReplication (45.36s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestEagerReplication PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestEagerReplication.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestEagerReplication failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2763813&tab=buildLog) on [master@7a8e7fb5e3a91ddcf752d30c068adcc88b8196d8](https://github.com/cockroachdb/cockroach/commits/7a8e7fb5e3a91ddcf752d30c068adcc88b8196d8):

```
=== RUN   TestEagerReplication
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestEagerReplication453158958
    test_log_scope.go:74: use -show-logs to present logs inline
    replicate_test.go:54: purgatory start count is 7
    replicate_test.go:68: condition failed to evaluate within 45s: expected 9 replicas in purgatory, but found 10
        goroutine 1544892 [running]:
        runtime/debug.Stack(0xc00222bb50, 0x54367a0, 0xc0103f4ec0)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x9f
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x5615ae0, 0xc0073e6000, 0xc00222bb50)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:36 +0x6d
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestEagerReplication(0xc0073e6000)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replicate_test.go:68 +0x4c5
        testing.tRunner(0xc0073e6000, 0x4e218e8)
        	/usr/local/go/src/testing/testing.go:1123 +0xef
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1168 +0x2b3
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestEagerReplication453158958
--- FAIL: TestEagerReplication (45.42s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestEagerReplication PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestEagerReplication.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/kv/kvserver/replicate_test.go
+++ b/pkg/kv/kvserver/replicate_test.go
@@ -15,7 +15,6 @@ import (
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
-	"github.com/cockroachdb/cockroach/pkg/kv"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/server"
@@ -33,6 +32,10 @@ func TestEagerReplication(t *testing.T) {
 
 	ctx := context.Background()
 	serv, _, _ := serverutils.StartServer(t, base.TestServerArgs{
+		// Need to trick the server to think it's part of a cluster, otherwise it
+		// will set the default zone config to require 1 replica and the split
+		// bellow will not trigger a replication attempt.
+		PartOfCluster: true,
 		Knobs: base.TestingKnobs{
 			Store: &kvserver.StoreTestingKnobs{
 				// Disable the replica scanner so that we rely on the eager replication code
@@ -46,6 +49,10 @@ func TestEagerReplication(t *testing.T) {
 	store, err := s.Stores().GetStore(s.GetFirstStoreID())
 	require.NoError(t, err)
 
+	// Make sure everything goes through the replicate queue, so the start count
+	// is accurate.
+	require.NoError(t, store.ForceReplicationScanAndProcess())
+
 	// After bootstrap, all of the system ranges should be present in replicate
 	// queue purgatory (because we only have a single store in the test and thus
 	// replication cannot succeed).
@@ -53,20 +60,15 @@ func TestEagerReplication(t *testing.T) {
 
 	t.Logf("purgatory start count is %d", purgatoryStartCount)
 	// Perform a split and check that there's one more range in the purgatory.
-
-	key := roachpb.Key("a")
-	args := adminSplitArgs(key)
-	_, pErr := kv.SendWrapped(ctx, store.TestSender(), args)
-	if pErr != nil {
-		t.Fatal(pErr)
-	}
+	_, _, err = s.SplitRange(roachpb.Key("a"))
+	require.NoError(t, err)
 
 	// The addition of replicas to the replicateQueue after a split
 	// occurs happens after the update of the descriptors in meta2
 	// leaving a tiny window of time in which the newly split replicas
 	// will not have been added to purgatory. Thus we loop.
 	testutils.SucceedsSoon(t, func() error {
-		expected := purgatoryStartCount + 2
+		expected := purgatoryStartCount + 1
 		if n := store.ReplicateQueuePurgatoryLength(); expected != n {
 			return errors.Errorf("expected %d replicas in purgatory, but found %d", expected, n)
 		}

